homeassistant/components/tado/config_flow.py

```diff
@@ -0,0 +1,148 @@
+"""Config flow for Tado integration."""
+import logging
+
+from PyTado.interface import Tado
+import requests.exceptions
+import voluptuous as vol
+
+from homeassistant import config_entries, core, exceptions
+from homeassistant.const import CONF_PASSWORD, CONF_USERNAME
+from homeassistant.core import callback
+
+from .const import CONF_FALLBACK, UNIQUE_ID
+from .const import DOMAIN  # pylint:disable=unused-import
+
+_LOGGER = logging.getLogger(__name__)
+
+DATA_SCHEMA = vol.Schema(
+    {vol.Required(CONF_USERNAME): str, vol.Required(CONF_PASSWORD): str}
+)
+
+
+async def validate_input(hass: core.HomeAssistant, data):
+    """Validate the user input allows us to connect.
+
+    Data has the keys from DATA_SCHEMA with values provided by the user.
+    """
+
+    try:
+        tado = await hass.async_add_executor_job(
+            Tado, data[CONF_USERNAME], data[CONF_PASSWORD]
+        )
+        tado_me = await hass.async_add_executor_job(tado.getMe)
+    except KeyError:
+        raise InvalidAuth
+    except RuntimeError:
+        raise CannotConnect
+    except requests.exceptions.HTTPError as ex:
+        if ex.response.status_code > 400 and ex.response.status_code < 500:
+            raise InvalidAuth
+        raise CannotConnect
+
+    if "homes" not in tado_me or len(tado_me["homes"]) == 0:
+        raise NoHomes
+
+    home = tado_me["homes"][0]
+    unique_id = home["id"]
+    name = home["name"]
+
+    return {"title": name, UNIQUE_ID: unique_id}
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Tado."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_CLOUD_POLL
+
+    async def async_step_user(self, user_input=None):
+        """Handle the initial step."""
+        errors = {}
+        if user_input is not None:
+            try:
+                validated = await validate_input(self.hass, user_input)
+            except CannotConnect:
+                errors["base"] = "cannot_connect"
+            except InvalidAuth:
+                errors["base"] = "invalid_auth"
+            except NoHomes:
+                errors["base"] = "no_homes"
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+
+            if "base" not in errors:
+                await self.async_set_unique_id(validated[UNIQUE_ID])
+                self._abort_if_unique_id_configured()
+                return self.async_create_entry(
+                    title=validated["title"], data=user_input
+                )
```

Comment:
For a future PR, we should really see if we can migrate this to use the OAuth2 endpoint so we don't need to store username/password in the config entry.

http://blog.scphillips.com/posts/2017/01/the-tado-api-v2/#oauth2